### ISSUE TITLE
[HOLD for payment 2023-11-27] [$1000] [P2P Activation] Chrome mWeb - scan button provides no feedback 

### ISSUE THREAD

**Issue by kevinksullivan** [Expensify/App#29990]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.3.85-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y

**Slack conversation:** https://expensify.slack.com/archives/C05RECHFBEW/p1697647198716039

## Action Performed:
1. Logged in on mWeb chrome
2. Selected `Scan`
3. Went through scan flow on this account and scanned a receipt via mobile web
4. Landed back in this flow
5. Tried to tap scan again

## Expected Result:
The camera should work as we already granted permission, OR we should minimally ask for permission again

## Actual Result:
I'm stuck on being unable to access the camera, and tapping the scan button doesn't prompt me to allow camera access or do anything

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [x] #30276
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c10a1b4074c90859</li>
        <li>Upwork Job ID: 1715136935013912576</li>
        <li>Last Price Increase: 2023-10-20</li>
<li>Automatic offers: </li>
<ul>
<li>Ollyws | Reviewer | 27295278</li>
<li>situchan | Contributor | 27326121</li>
</ul></li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
Unable to reproduce on Chrome, iOS. @kevinksullivan @Ollyws , how about you?

**Comment by mallenexpensify** [comment]
There are a few flows where I'm unable to reproduce but there is one where I am, it's something like. 
- On iOS Chrome, go to new.expensify.com and sign in. 
- Tap the green `+` sign in the bottom left corner (global create button)
- tap request money
- Tap `Scan` at the top, if needed. 
- Allow access (if needed) and scan photo
- after scanned, click the back button to go to LHN
- Refresh page (by clicking on the link in the address bar then tapping `go`
- Tap the green `+` sign in the bottom left corner (global create button)
- tap request money
- Observe that camera doesn't work

**Comment by mallenexpensify** [comment]
Also tagging @lukemorawski who might have insight here.

**Comment by alexander01202** [comment]
@mallenexpensify works for me. I followed all your steps

**Comment by alexander01202** [comment]
Contributor details
Your Expensify account email: frappes-topic0y@icloud.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f0e41431eb01dbdd

**Comment by alexander01202** [comment]
That's the email I'm currently logged in with

**Comment by alexander01202** [comment]
Contributor details
Your Expensify account email: lexiorient@icloud.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f0e41431eb01dbdd

**Comment by lukemorawski** [comment]
This a comment from Lucas of Callstack

**Comment by mallenexpensify** [comment]
Thanks for confirming my reproduction steps worked @alexander01202 . @Ollyws can you please test too? If you can also confirm I'll update the OP to be more details. 

I raised the price to $1k to get more eyes, since this is time sensitive.  
@jasperhuangg can you confirm this can be worked on externally?  (I'll remove your assignment after you confirm)

**Comment by b4s36t4** [comment]
@mallenexpensify as you mentioned ios chrome, does it means it only reproducible on physical device?

**Comment by trjExpensify** [comment]
@mallenexpensify we have a PR in review for this: https://expensify.slack.com/archives/C05RECHFBEW/p1697779669659709

**Comment by jasperhuangg** [comment]
@mallenexpensify Good shout! It looks like this can be worked on externally.

**Comment by trjExpensify** [comment]
Assigned @situchan as the C+ as he's been reviewing this.

**Comment by lukemorawski** [comment]
@Samueljh1 `navigator.permissions.query({ name: 'camera' })` will not ask for permissions. It'll just give you the current state of them (as name suggests).

**Comment by Samueljh1** [comment]
> @Samueljh1 `navigator.permissions.query({ name: 'camera' })` will not ask for permissions. It'll just give you the current state of them (as name suggests).

You are correct here. Internally, the permission is requested when calling `navigator.mediaDevices.getUserMedia({video:true})`. When doing the reproduction steps, the browser returns `Unhandled Promise Rejection: NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.`. No prompt is shown. Manually reloading the page causes this prompt to appear again, offering a workaround.

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

User is not able access the camera on Request money > Scan.
Tapping the Scan button again doesn't prompt user to allow camera access on Scan tab.

## What is the root cause of that problem?

**TLDR**: The camera permission prompt **does** show but due to a race condition of the react tab navigation the prompt is dismissed which automatically sets camera permission as **denied**.

https://github.com/Expensify/App/blob/e99c297332c738475cfe5701e0a00ba0b5d58029/src/pages/iou/ReceiptSelector/index.js#L194-L204

Issue comes from `ReceiptSelector` because due to the race condition that react tab navigation creates, the `NavigationAwareCamera` is rendered **before** the transition to the Scan tab is completed which **dismisses the camera permission prompt**. I've been able to reproduce this both on staging and local with current main code:

I added some slow-motion to the videos belod where the prompt can be seen briefly.
<details>
<summary>iOS mWeb: Chrome</summary>

- staging

https://github.com/Expensify/App/assets/56457735/f20b03d9-48e8-47a3-a300-d5e4ba6aa22a

- local

https://github.com/Expensify/App/assets/56457735/b46e1dcf-376b-4f70-8d17-7f85ca3fac47
</details>

The way I've been able to test locally is by being on the same wi-fi both on the mac and on a real iPhone device and setting up fully trusted root certificates for 192.168.1.135 (local ip) and adding the certs in `webpack.dev.config` before starting the app locally in order to be able to have secure `https://` locally on iPhone.

## What changes do you think we should make in order to solve the problem?

Unfortunately I couldn't find any other solution then using `setTimeout` for this edge case related to the react navigation which works because it was implemented in a similar fashion here:

https://github.com/Expensify/App/blob/e99c297332c738475cfe5701e0a00ba0b5d58029/src/components/TabSelector/TabSelector.js#L89-L94

So my working solution is the following:
- add a state variable called `isPermissionReady` which is set to initial value `!Browser.isMobile()`
- within a `useEffect` with `navigation.isFocused()` as dependency, check if (`Browser.isMobile()` && `navigation.isFocused()`) is true
- where a `setTimeout` with timeout of `CONST.ANIMATED_TRANSITION` (300ms) will call `setPermissionReady(true)`
- and in the `return` of the `useEffect` we set `setPermissionReady(false)` and we clear the timeout

By this we ensure that:
- if `Browser.isMobile()` is true, the `NavigationAwareCamera` is only rendered when `isPermissionReady` becomes true after the `navigation.isFocused()` and when the timeout is called
- this way the camera permission prompt is not displayed before the Scan tab is fully rendered (including animations)
- finally we make sure to set `isPermissionReady` back to false and clear the timeout when unmounting

### Caveats:
1. Once the camera works and user then decides to manually disable the camera permission, the camera will become black until user navigates between tabs then back to the Scan tab or performs browser refresh while on Scan tab in order for the permission prompt to be shown again and accepted.

<details>
<summary>iOS mWeb: Chrome - Caveat 1</summary>

https://github.com/Expensify/App/assets/56457735/c20ea0c6-2b57-4cdc-bf82-5decabd23c60

</details>

2. This implementation does not work if attempted from a desktop computer in the Chrome browser on an iOS simulator, or any mobile simulator for that matter.

<details>
<summary>MacOS - Desktop: Chrome mWeb Simulator - Screenshot</summary>

<img width="987" alt="Screenshot 2023-10-28 at 21 44 39" src="https://github.com/Expensify/App/assets/56457735/5691c2c9-553e-47a7-bb69-45fa180f5cf1">

</details>

### Code changes:
```diff
// src/pages/iou/ReceiptSelector/index.js

+    const [isPermissionReady, setPermissionReady] = useState(!Browser.isMobile());
    
+    useEffect(() => {
+        let timeoutID;
+        if (Browser.isMobile() && navigation.isFocused()) {
+            timeoutID = setTimeout(() => {
+                setPermissionReady(true);
+            }, CONST.ANIMATED_TRANSITION);
+        }

+        return () => {
+            setPermissionReady(false);
+            clearTimeout(timeoutID);
+        }
+    }, [navigation.isFocused()]);

+                {isPermissionReady && (
                    <NavigationAwareCamera
                        onUserMedia={() => setCameraPermissionState('granted')}
                        onUserMediaError={() => setCameraPermissionState('denied')}
                        style={{...styles.videoContainer, display: cameraPermissionState !== 'granted' ? 'none' : 'block'}}
                        ref={cameraRef}
                        screenshotFormat="image/png"
                        videoConstraints={{facingMode: {exact: 'environment'}}}
                        torchOn={isFlashLightOn}
                        onTorchAvailability={setIsTorchAvailable}
                        forceScreenshotSourceSize
                    />
+                )}
```

 ### Videos

<details>
<summary>iOS mWeb: Chrome</summary>

https://github.com/Expensify/App/assets/56457735/7955cd5a-bf7d-4731-ad83-624096c16457

</details>

**Comment by mallenexpensify** [comment]
@ikevin127 , someone is already hired for this job, sorry.  Please search for issues with the `Help Wanted` label. [This link should show all](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22). 

@lukemorawski , can you please address @situchan [comment in the PR?](https://github.com/Expensify/App/pull/30094#issuecomment-1783942374) 
> @lukemorawski please fix conflict when you get a chance

**Comment by lukemorawski** [comment]
working on some refactorings

**Comment by tgolen** [comment]
Reviews are wrapping up and I'm seeing about getting this merged today.

**Comment by kevinksullivan** [comment]
@Ollyws @situchan please fill out the checklist above when you get a chance!

**Comment by tgolen** [comment]
Bump @Ollyws and @situchan

**Comment by situchan** [comment]
I think the bug always existed after implementing scan feature.

We should create regression test for this since this bug is critical and main user flow.

### Regression Test Proposal (mWeb only)

1. Go to FAB > Request money
2. Tap Manual or Distance tab (not Scan)
3. Refresh app
4. Tap Scan tab
5. Verify that app asks for camera permission and grant permission (optional)
6. Verify that camera view shows

7. Click back button
8. Refresh app
9. Go to FAB > Request money
10. Verify that app asks for camera permission and grant permission (optional)
11. Verify that camera view shows

12. Refresh app
13. Verify that app asks for camera permission and grant permission (optional)
14. Verify that camera view shows

**Comment by mallenexpensify** [comment]
Contributor: @situchan paid $1000 via Upwork
Contributor+: @Ollyws owed $1000 via Upwork

@Ollyws can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01a0e882ab00ed78a6
(the old job closed so I created a new one)

Testrail update GH created
- https://github.com/Expensify/Expensify/issues/340448

**Comment by situchan** [comment]
@mallenexpensify we can close this.
[PR](https://github.com/Expensify/App/pull/30094) author is expert agency, C+ reviewer is just me

**Comment by mallenexpensify** [comment]
Ah... thanks @situchan , I didn't catch that, I just saw @Ollyws was assigned so I _assumed_ they were due payment.  

I updated the payment comment above to be accurate.
